pkg/storegateway/indexheader/index/postings.go

```diff
@@ -373,25 +373,55 @@ func (t *PostingOffsetTableV2) PostingsOffset(name string, value string) (r inde
 	return index.Range{}, false, nil
 }
 
-func (t *PostingOffsetTableV2) LabelValues(name string, filter func(string) bool) (v []string, err error) {
+func (t *PostingOffsetTableV2) LabelValues(name string, prefix string, filter func(string) bool) (v []string, err error) {
 	e, ok := t.postings[name]
 	if !ok {
 		return nil, nil
 	}
 	if len(e.offsets) == 0 {
 		return nil, nil
 	}
-	values := make([]string, 0, len(e.offsets)*t.postingOffsetsInMemSampling)
+
+	offsetIdx := 0
+	if prefix != "" {
```

Comment:
This logic might be a good candidate for its own method. WDYT?

```go
func (t *PostingOffsetTableV2) findClosestOffsetIdx(prefix string) (index int, ok bool)
```

Reply:
Agree, especially since the method is the same for `PostingOffsetTableV2` and for `BinaryReader`.

Reply:
Done in https://github.com/grafana/mimir/pull/4055/commits/18ecccc169aec596abacd01175e015507f17e5e0

I didn't try to export stuff to reuse them from BinaryReader as I assume that will be deleted anyway soon.

Reply:
:crossed_fingers: